system/lib/pthread/pthread_join.c

```diff
@@ -28,17 +30,39 @@ static int __pthread_join_internal(pthread_t t, void **res, int tryjoin) {
     // thread is attempting to join to itself.
     return EDEADLK;
   }
-  return __pthread_join_js(t, res, tryjoin);
+  int is_main_thread = emscripten_is_main_runtime_thread();
+  while (1) {
+    // The thread we are joining with must be either DT_JOINABLE or
+    // DT_EXITING.  If its DT_EXITING then we move it to DT_EXITED and
+    // we are done.   If its DT_JOINABLE we keep waiting.
+    int old_state = a_cas(&t->detach_state, DT_EXITING, DT_EXITED);
+    if (old_state == DT_EXITING) {
+      // We successfully marked the tread as DT_EXITED
+      if (res) *res = t->result;
```

Comment:
This load was atomic in JS before?

Reply:
Hmm.. I would that here is not a need an explicit amount read here.   I'm hoping we can just match the upstream musl  code:

https://github.com/emscripten-core/emscripten/blob/626692fb71ad8836157ed9c0b2bfd74b37f23ca3/system/lib/libc/musl/src/thread/pthread_join.c#L23

Is the synchronization on the detach_data above via the `a_cas` not sufficient?   The result is wrong during exit (also not using an explict atomic call) before that `a_cas` is used to signal the join is ready.

https://github.com/emscripten-core/emscripten/blob/626692fb71ad8836157ed9c0b2bfd74b37f23ca3/system/lib/pthread/pthread_create.c#L193

This swap is done only after writing the result above.

musl doesn't seem to have the concept an atomic read and seems to volatile along with `a_cas` and other primitives.  They have a whole bunch of atomic ops but basic read and write are not among them:
https://github.com/emscripten-core/emscripten/blob/main/system/lib/libc/musl/src/internal/atomic.h

It would feel a little odd to add `a_read` and `a_write` when musl doesn't do this.. (or use to emscripten_atomic_read here).

Reply:
I agree that we could revisit the use of bare reads/write in places like this, and I'm happy audit all the places where musl seems to assume that a_cas is a memory barrier.   

For now I'd like to land this since it passes all the tests including all the posixtestsuite.